### ISSUE TITLE
R4: Integration validation and handoff Implementation Plan

### ISSUE THREAD

**Issue by JueonPark** [JueonPark/sym#148]
**Goal:** Make the completed compiler-to-runtime path reproducible from a fresh checkout and publish evidence for supported input, output, dynamic, typed, and weight-loading scenarios.

**Architecture:** Integrate R1–R3 with T3/T4 and C4 through runnable public examples and a shared acceptance matrix. Reuse their APIs and examples, add a small standalone C++ artifact consumer, and make CI/evidence distinguish CPU conformance from actual CUDA transport. Keep performance descriptive and historical claims explicit.

**Tech Stack:** LLVM/MLIR 21.1.8 compiler, standalone C++ runtime, CUDA 12.6, pybind11 3.0.4, regular-GIL CPython 3.14.7 and PyTorch 2.14.0 CPU/cu126. Use the selected baseline's separate CPU/CUDA environments and matching extension builds.

**Spec:** [#133](https://github.com/JueonPark/sym/issues/133), [project plan §3/R4 and final handoff](https://github.com/JueonPark/sym/blob/main/docs/project-finalization-plan.md#3-complete-compiler-to-runtime-execution), [T4 dynamic/weight plan](https://github.com/JueonPark/sym/blob/main/docs/torch-finalization/t4-dynamic-inputs-and-weights.md), [C4 conformance](https://github.com/JueonPark/sym/issues/144), and [claim ledger](https://github.com/JueonPark/sym/blob/main/docs/claim-ledger.md).

## Global Constraints

- "Keep torch dependencies outside the core runtime."
- "Unsupported cases execute through the original PyTorch path with a recorded reason."
- Supported examples use generated plans and public APIs, without diagnostic scraping or benchmark-private setup.
- A fresh checkout must reproduce documented supported behavior; missing prerequisites must fail a required validation leg rather than silently skip it.
- CPU evidence does not certify CUDA stream/lifetime correctness. Record the actual device, wheel, toolkit, extension, compiler revision, command, and outcome.
- Performance is reported descriptively, without a speedup gate. Preserve historical failed/withdrawn claims and separate research dispositions from functional acceptance.

## Review Focus

- Examples can accidentally work only because of stale artifacts, an old extension, or untracked local environment state (Task 1).
- A matrix can claim D2H correctness using only inverse round-trip checks (Task 2).
- Portable artifacts and dynamic caches can remain loadable while bindings or weights are stale (Task 2).
- CUDA skips or mocked dispatch can make an integration job appear to validate transfers (Task 3).
- Closing historical issues can incorrectly imply their original performance conditions were achieved (Task 4).

---

**Proposed child title:** `[Finalize][Runtime][R4] Validate end-to-end execution and publish the handoff`

**Dependencies:** [R1](https://github.com/JueonPark/sym/issues/145)–[R3](https://github.com/JueonPark/sym/issues/147); [T3 / #136](https://github.com/JueonPark/sym/issues/136), [T4 / #137](https://github.com/JueonPark/sym/issues/137), and [C4 / #144](https://github.com/JueonPark/sym/issues/144) supply the frontend, dynamic/weight lifecycle, and typed conformance surfaces. Prepare the matrix and build instructions early; close this child only after combined execution evidence is available. [R4](https://github.com/JueonPark/sym/issues/148) does not introduce another importer, dispatcher, weight cache, or numerical oracle.

**Runtime subissues:** [R1](https://github.com/JueonPark/sym/issues/145), [R2](https://github.com/JueonPark/sym/issues/146), [R3](https://github.com/JueonPark/sym/issues/147), [R4](https://github.com/JueonPark/sym/issues/148).

## Starting evidence

Main `4e67545d2e6331802a7e295c65c135cf90b26a4f` delivers T1/T2 and R1's exporter/artifact prerequisite. Its support documentation explicitly says transfer replacement is not enabled. Existing CPU/CUDA inventory evidence is useful baseline context, but it is not acceptance evidence for the future R2/R3/T3 execution path. T4 already plans dynamic-input and weight-loading examples; C4 plans the typed relocation example and numerical support matrix.

## File map

| Action | Path | Responsibility |
| --- | --- | --- |
| Create | `docs/runtime-integration.md` | One entry point for build, public C++/Python usage, combined support matrix, evidence and reproduction |
| Create | `libreloc/examples/RunArtifact.cpp`, `libreloc/examples/CMakeLists.txt` | Small standalone C++ consumer of public compiler-generated wire bytes |
| Modify | `libreloc/CMakeLists.txt` | Register the example target without adding compiler/Torch dependencies to the runtime consumer |
| Create | `libreloc/python/examples/compiler_runtime_handoff.py` | Reproducible coordinator for the accepted examples and machine-readable results |
| Reuse/extend | `libreloc/python/examples/torch_dynamic_transfers.py`, `libreloc/python/examples/torch_weight_loading.py` (T4) | Input/output, symbolic reuse, module parameter/buffer and weight lifecycle examples |
| Reuse/extend | `libreloc/python/examples/torch_typed_relocation.py`, `docs/typed-relocation-support.md` (C4) | Mixed typed examples and authoritative numerical support rows |
| Create | `libreloc/python/tests/torch_frontend/test_runtime_integration.py` | Combined generated-artifact/public-entrypoint regressions and runner accounting |
| Modify | `.github/workflows/build.yml`, `libreloc/test/CMakeLists.txt` | Require compiler/runtime integration tests and example smoke runs |
| Create | `docs/runtime-evidence/cpu.json`, `docs/runtime-evidence/cuda.json` | Compact reproducible integration results with versions and actual execution/fallback evidence |
| Modify | `README.md`, `libreloc/README.md`, `docs/torch-support.md`, `docs/torch-integration.md` (T4), `docs/torch-finalization/README.md` | Link the entry point and reconcile current support/status statements |
| Modify | `docs/project-finalization-plan.md`, `docs/claim-ledger.md` | Final acceptance references and explicit historical issue dispositions |

## Task 1: Reproduce public artifact execution from a clean build

**Interfaces:** Add `reloc-run-artifact PLAN --symbols NAME=VALUE,... --input INPUT --output OUTPUT --direction host|h2d|d2h`. Inputs/outputs are raw declared scalar storage described by the companion documented recipe; symbol names must match the wire table. The example uses public decode/bind/transfer interfaces, validates sizes, reports errors nonzero, and never parses compiler diagnostic output. Its C++ target links the runtime and optional CUDA support, not MLIR/LLVM/Torch.

- [ ] Document exact compiler and separate runtime/Python build commands from the existing CI configuration, including exporter path, matching cp314 extension path, dependency pins, CPU/cu126 wheel selection, and CUDA architecture/toolkit settings. Use fresh build/artifact directories and record resolved versions. Do not overwrite the user's existing environment as part of reproduction.
- [ ] Add checked-in literal layout recipes or reuse existing canonical exporter fixtures to generate the C++ example input artifacts. Exercise one host layout execution and, on CUDA, forward H2D/D2H via R2. Supply deterministic raw input generation and an independent output comparison command in the guide. Validate unknown symbols, short buffers, and malformed artifacts before execution.
- [ ] Make the example own host/device allocations and wait before writing output. For H2D validation, copy the final device result back solely to produce the comparison file; report that verification readback separately from the requested transfer. For D2H validation, stage the documented logical source on the device before invoking the requested forward recipe. The C++ example need only demonstrate the v0 layout public path; C4's Python example owns mixed typed demonstration.
- [ ] Reuse R1's portable `CompiledRecipe` save/load path in a fresh Python subprocess with concrete symbol rebinding. After C3 lands, cover new-runtime v0/v1 loading and explicit old-runtime rejection of unsupported typed versions. Avoid inventing another artifact container.
- [ ] Create `compiler_runtime_handoff.py --device cpu|cuda --output PATH` to invoke the documented accepted examples and write JSON scenario results. Validate prerequisites up front: a requested CUDA run with no GPU/CUDA runtime is an error; expected unsupported scenarios are reported with their reasons rather than counted as successful offloads.
- [ ] Run the new runner from a fresh configured checkout using only tracked files and documented commands. Add focused tests for missing exporter/extension, malformed result files, and failed child commands; the coordinator must return nonzero for a required failed scenario.
- [ ] Commit as `docs(runtime): make public artifact examples reproducible`.

## Task 2: Close the combined behavior matrix

**Interfaces:** Each integration row records recipe, direction/dtypes, shape cases, expected result/metadata, required implementation or fallback reason, and a reproducible test/example reference. Consume R3's actual dispatch report and T3/T4's scalar-only diagnostics; do not retain tensors in evidence records.

- [ ] Cover layout identity, reshape/transpose, constant pad, and exact f32/f16/int8 bytes. Rebind a single symbolic artifact at 64/128/192 and assert expected compile/load/bind behavior using T4's counters; check guard-invalid shapes take the original path once.
- [ ] Cover input H2D and output forward D2H with a non-self-inverse permutation and a padding witness, checking logical shape/strides/device/dtype and fresh storage. Reuse R2's default/non-default stream, immediate-consumer, and allocation-reuse tests as named prerequisites, then exercise the same path through T3's public entry point.
- [ ] Run C4's mixed layout+cast and layout+quantize/dequantize examples through R3. Record forced `original_cpu` and auto implementation reports, exact source/wire/destination/parameter footprints, and comparisons under C1's numerical rules. Show an unsupported optimized kernel correctly using a qualified original path or a reason-coded frontend fallback.
- [ ] Include invalid/malformed/stale artifacts, missing calibration, explicit nonblocking exclusions, unsupported empty/rank/offset/overlap cases, and transfer failure propagation. Assert guard fallback invokes the original callable once; injected execution errors must not trigger a second execution.
- [ ] Reuse T4's module parameter/buffer loading, dynamic reuse, live slot ownership, `load_state_dict`, in-place mutation, storage replacement, parameter changes, close, and collection scenarios. Verify prepared/prefolded weights invalidate when values or relevant configuration change, including retained objects with unchanged addresses.
- [ ] Test runtime resource cleanup after successful calls and failures. Exercise backend/prefold owner destruction order through the public integration, not merely native unit cases. Record unavailable multi-GPU/hardware configurations as unqualified support boundaries.
- [ ] Add only missing combined regressions to `test_runtime_integration.py`; keep the numerical oracle and exhaustive component cases in their existing owners. Commit as `test(runtime): validate generated plans through integrated frontend paths`.

## Task 3: Require CPU CI and collect real CUDA acceptance evidence

**Interfaces:** `docs/runtime-evidence/{cpu,cuda}.json` records source revision, exact build/runtime versions and extension path, commands, scenario outcomes, dispatch/fallback evidence, and measured byte counts. Compact evidence must be generated by the runner or documented commands, not handwritten success claims.

- [ ] Extend the existing CPU cp314 job to explicitly import the intended extension, require the real exporter, run CTest/lit and CPU integration/reference tests, and smoke-test the public examples. Preserve runtime symbol/include/dynamic-dependency checks and the separate Torch-free runtime environment.
- [ ] Run CUDA integration on actual hardware with a CUDA-enabled runtime and the qualified cu126 environment. Require executed GPU cases for default/non-default streams, forward H2D/D2H, typed kernels, immediate consumers, repeated allocation, and applicable weight/prefold execution. A GPU test selection with zero executed tests is failed acceptance.
- [ ] Save environment and command outcomes in the evidence files and link reproducible CI runs or recorded hardware runs. Separate previous T1 inventory evidence from new execution evidence; avoid labeling untested architectures or wheel variants supported.
- [ ] Measure descriptive end-to-end latency and actual bytes for original and selected implementations, recording warmup, synchronization boundaries, shape/reuse parameters, and whether compile/preparation time is included. Do not add a speedup threshold or modify the central cost model to make the examples pass.
- [ ] Verify the complete documented commands from a fresh checkout and commit as `ci(runtime): require integration evidence and runnable examples`.

## Task 4: Publish the support boundary and project handoff

**Interfaces:** `docs/runtime-integration.md` is the navigation/acceptance entry point; Torch and typed support documents remain authoritative for their detailed rows. The project plan and claim ledger link implementation evidence and historical dispositions.

- [ ] Reconcile obsolete status text, including T1/T2 issue states and R1 delivery, with actual completed work. Document supported layouts/dtypes/directions, artifact compatibility, blocking semantics, parameter ownership, fallback behavior, and explicit exclusions. Link T4/C4 examples instead of maintaining divergent copies.
- [ ] Publish named input, output, and weight-loading commands and expected metadata/results; give users a direct way to force R3's original CPU pipeline and inspect the selected path without benchmark infrastructure.
- [ ] Review [#55](https://github.com/JueonPark/sym/issues/55) and [#56](https://github.com/JueonPark/sym/issues/56) against decoder/pad coverage; [#57](https://github.com/JueonPark/sym/issues/57) and [#71](https://github.com/JueonPark/sym/issues/71) for separate refactoring/sanitizer disposition; and [#88](https://github.com/JueonPark/sym/issues/88) for the applicable weight-loading/overlap scenario. Record evidence or remaining work for each rather than automatically claiming its original acceptance is satisfied.
- [ ] Preserve failed/withdrawn research claims for [#63](https://github.com/JueonPark/sym/issues/63) and [#73](https://github.com/JueonPark/sym/issues/73) in the claim ledger. Any proposed historical closure must state concluded/superseded scope where appropriate. Keep unpublished `human_history/` records local unless separately authorized for publication.
- [ ] Record whether each parent meets its functional acceptance and list any remaining exclusions or historical decisions. Close R4 only when required CPU and actual CUDA integration evidence is linked; do not use benchmark performance as the completion gate.
- [ ] Commit as `docs(runtime): publish final support evidence and project disposition`.

## Acceptance and handoff

- [ ] A fresh checkout reproduces public compiler→artifact→load/bind→execute paths from Python and C++ without benchmark-private setup.
- [ ] Input/output/weight examples, symbolic reuse, v0/v1 compatibility, typed conversions, fallback, cleanup, and prefold invalidation have named results.
- [ ] CPU CI requires compilation/binding/reference execution; actual CUDA evidence proves transfers, stream ordering, and lifetime behavior.
- [ ] Support documents agree with executed capability; unsupported configurations retain defined behavior.
- [ ] Performance remains descriptive and historical issue dispositions are explicit, with no unsupported claims of research success.

Once Tasks 1–3 register the examples, run with `PYTHONPATH`/`SYM_RELOC_EXPORT` selecting the matching configured build:

```bash
ctest --test-dir build/sym --output-on-failure
.venv/bin/python -m pytest libreloc/python/tests/torch_frontend/test_runtime_integration.py -m 'not gpu' -q
.venv/bin/python libreloc/python/examples/compiler_runtime_handoff.py --device cpu --output /tmp/runtime-handoff-cpu.json
# On the qualified CUDA environment with a CUDA-enabled runtime:
.venv/bin/python -m pytest libreloc/python/tests/torch_frontend/test_runtime_integration.py -m gpu -q
.venv/bin/python libreloc/python/examples/compiler_runtime_handoff.py --device cuda --output /tmp/runtime-handoff-cuda.json
```
